### PR TITLE
Evaluation service integration

### PR DESCRIPTION
**Pull request by christophertubbs**
Migration of evaluation service GUI functionality into the primary MaaS gui and the adoption of basic websocket handlers in the dmod.communication pattern.

### CODE REVIEW

**Review by aaraney — CHANGES_REQUESTED** [review]
Nice work! I got a chance to review 1/3 of this and i'v left some questions and comments. Most of it is all minor, just type hinting stuff and minor questions. Looking forward to working through it! Thanks!

File: python/gui/MaaS/cbv/evaluation.py
Comment (christophertubbs) [reply python/gui/MaaS/cbv/evaluation.py:27]: The solution here isn't that important as long as it's somewhat memorable. What would be nice would be knowing the user's time zone - that way you could just know that you ran something at around 10:50am, and don't have to add 5 to it to get 15:50. Something like `{random adjective}-{random noun}` would work as well.

Either way, this wouldn't really be the core entrypoint - just a "See! It works!" entrypoint, if that makes any sense.

Can we add that as a new ticket or would you prefer to have it fixed prior to approval?

File: python/gui/maas_experiment/application_values.py
Comment (christophertubbs) [reply python/gui/maas_experiment/application_values.py:220]: No clue. A lot of that is just a straight copy and paste from work committed around a year ago. I'd prefer a hard default, but I don't know what it should be. I *think* L201 ensures that `MAAS_ENDPOINT_PORT` is always defined, though.

File: python/gui/maas_experiment/application_values.py
Comment (christophertubbs) [reply python/gui/maas_experiment/application_values.py:261]: `host == host_config['host']` due to line 239. I'm removing that line and just referencing `host_config['host']` so it'll be less misleading.

File: python/lib/communication/dmod/communication/evaluation_request.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/evaluation_request.py:43]: I *think* so? I don't see *where* it might lead things off the rails. Should I switch it up prior to acceptance or should we just keep an eye out for it and any possible similar situations?

File: python/lib/communication/dmod/communication/evaluation_request.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/evaluation_request.py:51]: That threw me off, too: the deserializer basic rolls through a list of approved request classes that all have this function. If it successfully deserializes, it is interpreted as being the type of class to use, which then indicates the type of handler to use, so that's just part of the infrastructure at this point in the request service. I'll go ahead and add the annotation.

File: python/lib/communication/dmod/communication/evaluation_request.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/evaluation_request.py:93]: I looked at making it a property but I think there was a need somewhere to lift this right off the type and not an instance.

File: python/lib/communication/dmod/communication/message.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/message.py:54]: I'll give you a dollar if you can tell me why I did it that way. The change is incoming.

File: python/lib/communication/dmod/communication/registered/aliases.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/registered/aliases.py:1]: Good catch. The file originally had ~20 aliases, but they ended up getting factored out of the equation. It's probably a matter of time before the file itself is removed. Future push will have those removed.

File: python/lib/communication/dmod/communication/message.py
Comment (aaraney) [reply python/lib/communication/dmod/communication/message.py]: 🤣

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:62]: Print may have just been for debugging and just stayed in. Not sure. I don't think all the logs are going to stdout and that entire mechanism is a tad scattered for now.

File: python/gui/MaaS/cbv/evaluation.py
Comment (aaraney) [reply python/gui/MaaS/cbv/evaluation.py:27]: I get what you mean. Yeah, we might want to think about "standardizing" evaluation id's in the future, but for this is good with me. Ill open an issue so I dont forget about this.

File: python/gui/maas_experiment/application_values.py
Comment (aaraney) [reply python/gui/maas_experiment/application_values.py:220]: Didnt make the connection between the two. Makes sense, thanks for clarifying.

File: python/gui/maas_experiment/settings.py
Comment (aaraney) [inline python/gui/maas_experiment/settings.py:128]: Does `django` magically import `CURRENT_TIMEZONE` from somewhere else (im still new to django settings)? I don't see `CURRENT_TIMEZONE` defined or imported elsewhere.

File: python/lib/communication/dmod/communication/evaluation_request.py
Comment (aaraney) [reply python/lib/communication/dmod/communication/evaluation_request.py:43]: I think it is fine. Given how classmethods are used elsewhere in the codebase, I think I am just getting use to how we are using them.

File: python/lib/communication/dmod/communication/message.py
Comment (aaraney) [inline python/lib/communication/dmod/communication/message.py:56]: Thanks!

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (aaraney) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:62]: Can you log this instead, please sir?

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (aaraney) [inline python/lib/communication/dmod/communication/registered/registered_interface.py:106]: Thanks for all the explicit type hints!

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (aaraney) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:217]: For sure!

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (aaraney) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:261]: Hahaha, long-term durability makes sense.

File: python/gui/maas_experiment/settings.py
Comment (christophertubbs) [reply python/gui/maas_experiment/settings.py:128]: `CURRENT_TIMEZONE` is coming in from `application_values`. It's defined there because you have to spin up another server instance to use the stuff in `settings.py`, which make cross-process stuff a problem. It's good to have everything in one, safe place.

File: python/lib/core/dmod/core/common/__init__.py
Comment (aaraney) [inline python/lib/core/dmod/core/common/__init__.py:6]: Would you consider adding `__all__` to `helper_functions` or manually importing each function here to avoid clouting the namespace with imports in `helper_functions`?

File: python/lib/core/dmod/core/common/failure.py
Comment (aaraney) [inline python/lib/core/dmod/core/common/failure.py:27]: I really like this method name, it made me chuckle.

File: python/lib/core/dmod/core/common/tasks.py
Comment (aaraney) [inline python/lib/core/dmod/core/common/tasks.py:95]: What are your thoughts @christophertubbs on returning a boolean that indicates success from this coroutine?

File: python/lib/core/dmod/core/common/tasks.py
Comment (aaraney) [inline python/lib/core/dmod/core/common/tasks.py:157]: ditto, you could return a tuple of booleans that indicate the success or failure of cancellation. That might not make as much sense here given that you could pass in an unordered iterable of `tasks`.

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (aaraney) [inline python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:684]: Thoughts on changing this to a `TypeError`?

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (aaraney) [inline python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:732]: I was kind of thinking the same thing.

Comment (aaraney) [reply python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:732]: Also, if we are awaiting here, that means we cannot receive any other messages, correct? We might want to consider how to avoid blocking incoming messages while we handle a received message.

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (christophertubbs) [reply python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:684]: There's a bunch of places where TypeError makes more sense. Changing it now.

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (christophertubbs) [reply python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:732]: We'll need to revisit the design. I'm going to add a todo because you have a good point, but it's easier said than done - there's already two threads associated with each duplex handler, so we don't want to be too naive with creating new groups of threads. What _might_ work would be to have an intermediary map that maps handler names to the async results. Once all have been launched, await the group. Once the group is done, assign the results to the response data.

Is it ok if we put it on the back burner for now? These external request handlers are just playing hot potato in their current state.

File: python/lib/metrics/__init__.py
Comment (aaraney) [inline python/lib/metrics/__init__.py:7]: This is for sure not critical, (I am assuming this is code re-use) we can get rid of `object`.

Comment (aaraney) [reply python/lib/metrics/__init__.py:7]: Also, the shebang since this will not be run from the command line directly.

Comment (aaraney) [reply python/lib/metrics/__init__.py:7]: Ive noticed that several places, but just have not commented on it. I think it makes the most sense to open an issue to clean unnecessary shebangs in the future.

File: python/lib/metrics/__init__.py
Comment (aaraney) [inline python/lib/metrics/__init__.py:48]: Definitely not major, but we may want to move the main function into `__main__.py` so it can be run using `python -m dmod.metrics`.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html
Comment (aaraney) [inline python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html:125]: For the purposes of this PR, I think this (inputs / form) is perfectly fine and what we need. However, in the future, we might consider using [`django` forms](https://docs.djangoproject.com/en/4.1/topics/forms/#building-a-form-in-django) in the future.

File: python/lib/core/dmod/core/common/tasks.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/tasks.py:95]: I'm having it return a new `CancelResults` type. That way it'll give that boolean _and_ say what failed to cancel. That function is used by `cancel_tasks`, so I imagine that should return something as well, but a tuple of booleans isn't particularly enlightening.

File: python/lib/core/dmod/core/common/tasks.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/tasks.py:157]: See above; it's going to return a tuple of results stating what the task name was, if it was successfully ended, and any attached messaging.

**Review by aaraney — DISMISSED** [review]
@christophertubbs, I just completed a full review. Overall, it looks really solid! Ive left a few comments that are mainly either stylistic or easily addressable in the future. I would still like it if @robertbartel had a chance to review this before we merge, but either way, great stuff!

File: python/gui/maas_experiment/settings.py
Comment (aaraney) [reply python/gui/maas_experiment/settings.py:128]: Duh, my bad. Didnt notice the wildcard import at the top! Sorry!

File: python/lib/metrics/__init__.py
Comment (christophertubbs) [reply python/lib/metrics/__init__.py:48]: That whole file is left over from an old template. I'm going to wipe the file to get all of that out of it, but it's questionable whether or not there should be an init there - it might end up conflicting with other inits when building and that's not something I'm prepared to try to break and fix.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html:125]: _Luckily_, this view shouldn't be used that often (if ever). See `$("#launch-warning")` as to why. Also, I seem to really hate django form rendering for whatever reason. If that's what you've been using for some of the new pages in the main GUI this might be a good place for me to practice the new style.

File: python/lib/core/dmod/core/common/__init__.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/__init__.py:6]: I'm going to go with the individual imports.

File: python/lib/core/dmod/core/common/tasks.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/tasks.py:95]: Hmm, im not seeing where `CancelResults` is being returned? Did that get committed?

File: python/lib/core/dmod/core/common/__init__.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/__init__.py:6]: Thanks!

File: python/lib/core/dmod/core/common/tasks.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/tasks.py:157]: Right, so like above, `cancel_tasks` just needs to return the `await asyncio.gather`

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (aaraney) [reply python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:684]: Tanks!

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:106]: That particular type hint will need to change in the future. I don't remember what version it is, but `typing` is going to support generic variable arguments (`*args` and `**kwargs`) at some point in the future. I think we have a ways to go, so this is just a hack.

File: python/lib/externalrequests/dmod/externalrequests/duplex/handler.py
Comment (aaraney) [reply python/lib/externalrequests/dmod/externalrequests/duplex/handler.py:732]: Yeah, I completely agree about back burnering this for now. Whatever we decide will definitely add some complexity like you are alluding to.

File: python/lib/communication/dmod/communication/registered/registered_interface.py
Comment (christophertubbs) [reply python/lib/communication/dmod/communication/registered/registered_interface.py:62]: logging on the way.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html
Comment (aaraney) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/templates/evaluation_service/ready_evaluation_async.html:125]: Thanks for the call out, ill give that a read. Yeah, to be fair, django forms are kind of wonky if you want to add js validation or do anything mildly complicated. I think they are only usable if you are cool styling them with `flexbox` or bet yet `grid`. I am using them in the [dataset manager view](https://github.com/NOAA-OWP/DMOD/blob/085dab4c534000b22c5c2924ba0df6c2b89528c2/python/gui/MaaS/cbv/DatasetManagementForms.py) for the at the moment.

**Review by aaraney — APPROVED** [review]

**Pull request merged**
